Add new services to alarm_control_panel

jnimmo opened the pull request:
**Description:**
Added two new services to the alarm_control_panel required for Envisalink functionality

alarm_keypress allows a string of custom keypresses to be sent to the alarm 
alarm_output_control activates the specified programmable output on the alarm

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#1249

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

Commit:
Merge remote-tracking branch 'home-assistant/dev' into add-methods-to-alarm-control-panel

Commit:
Merge remote-tracking branch 'home-assistant/dev' into add-methods-to-alarm-control-panel

Commit:
Add two new alarm services

Send custom keypresses to an alarm, or control the programmable outputs
on an alarm with the alarm_keypress and alarm_output_control services

Commit:
Reference new pyenvisalink release

Commit:
Updated requirements_all.txt

Commit:
Tidy up for flake8 & pylint

Commit:
pydocstyle changes

Cinntax commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -52,13 +59,24 @@ def setup(hass, config):
-
         code = service.data.get(ATTR_CODE)
 
-        method = SERVICE_TO_METHOD[service.service]
```
I think you need to keep this, as well as the getattr below. Then I'd look at a simple if statement like so:
If service.service == SERVICE_ALARM_KEYPRESS:
    ...get the keys to press from service data
    ...pass the result of that to getattr instead of "code"
....same for output control
....else:
    ...get the code value
    ...call the getattr line just like it is now

Make sense?  Sorry- on iPhone typing this.

jnimmo replied:
@Cinntax Ok could do, did think it might not be OK the way I had done - just though it made code clearer to understand what was happening with the if statements
@MartinHjelmare Cover component seems to handle it well - I don't understand how it it passes the appropriate parameters though? Any tips for understanding that?
When I tried it like that in Envisalink it kept passing code to the keypress parameter which wasn't helpful.

MartinHjelmare replied:
You should create different service schemas for the different services, since your new services have required arguments. The schemas will validate the parameters/attributes and will error before calling the method if a wrong parameter is included.

Include the schemas in the dictionary mapping service to method, like in the cover component. This map will also make sure that only the method connected to the service is called. Look up the dictionary for the correct schema when registering the services.

Commit:
Updated service handler to match cover component

Setup schemas to handle the different parameters

Commit:
Merge remote-tracking branch 'home-assistant/dev' into add-methods-to-alarm-control-panel

jnimmo commented:
Thanks very much I've replicated the way the cover component handles it and implemented different schemas for the new services. Tox happy and alarm services are working as expected

MartinHjelmare commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -21,26 +22,40 @@
+})
+
+ALARM_KEYPRESS_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
```
This should be optional.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -21,26 +22,40 @@
+})
+
+ALARM_OUTPUT_CONTROL_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
```
This should be optional.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -50,25 +65,26 @@ def setup(hass, config):
     def alarm_service_handler(service):
-        """Map services to methods on Alarm."""
-        target_alarms = component.extract_from_service(service)
+        """Handle calls to the cover services."""
```
Update name of component.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -116,6 +132,28 @@ def alarm_trigger(hass, code=None, entity_id=None):
+def alarm_keypress(hass, keypress, entity_id=None):
+    """Send a custom key sequence to the alarm."""
+    data = {}
+    if keypress:
```
`keypress` is a positional argument, so it will have to be passed when calling this function. You should assume it exists.

jnimmo replied:
Just to confirm I just need to remove the if statement for the keypress here?

MartinHjelmare replied:
Yes.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -116,6 +132,28 @@ def alarm_trigger(hass, code=None, entity_id=None):
+def alarm_output_control(hass, output, entity_id=None):
+    """Toggle an output on the alarm."""
+    data = {}
+    if output:
```
Same as above.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -146,6 +184,14 @@ def alarm_trigger(self, code=None):
 
+    def alarm_keypress(self, keypress=None):
+        """Send custom key sequence to alarm."""
+        raise NotImplementedError()
```
Are you sure this should be a required method for all alarms? I suggest replacing this with `pass`.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -146,6 +184,14 @@ def alarm_trigger(self, code=None):
+
+    def alarm_output_control(self, output=None):
+        """Control an output on the alarm."""
+        raise NotImplementedError()
```
Same as above.

MartinHjelmare commented on homeassistant/components/frontend/www_static/home-assistant-polymer:
```diff
@@ -1 +1 @@
-Subproject commit 75d760fc3e6f37485ae535a5476c8b00ecd2dfec
+Subproject commit db109f5dda043182a7e9647b161851e83be9b91e
```
I don't think this change should be included.

jnimmo replied:
Can't figure out how to get rid of this.. should I just delete this branch/PR and start a new one just with the 5 changed files? Git checkout and a hard reset does nothing

MartinHjelmare replied:
Checkout dev branch, pull from upstream, update submodule, checkout your feature branch, squash your commits, rebase off dev branch. That should do it.

The commit in the submodule you've changed is the latest in the dev branch, so you just need to make sure you are based off latest dev.

jnimmo replied:
Hi @MartinHjelmare are you able to help review PR 3932 at all? It seems to have become stuck! Thanks

MartinHjelmare commented on homeassistant/components/alarm_control_panel/services.yaml:
```diff
@@ -41,3 +41,25 @@ alarm_trigger:
+      example: 'alarm_control_panel.downstairs'
+    output:
+      description: Output number on the alarm to toggle
+      example: 1
```
Needs an additional newline after the last character in the file.

MartinHjelmare reviewed (changes requested):
Good. Some changes needed.

jnimmo committed:
Tidying up for pull request

Make entity_ids optional, fix commenting and remove unneeded checks for parameters existing

jnimmo committed:
Add newline

Commit:
Merge remote-tracking branch 'home-assistant/dev' into add-methods-to-alarm-control-panel

jnimmo commented:
Sorry I wasn't able to tidy up this branch, I've re-forked and made the changes and put into [PR3932](https://github.com/home-assistant/home-assistant/pull/3932)
